src/boot/efi/pe.c

```diff
@@ -239,15 +240,58 @@ EFI_STATUS pe_kernel_info(const void *base, uint32_t *ret_compat_address) {
         return EFI_SUCCESS;
 }
 
-EFI_STATUS pe_memory_locate_sections(const void *base, const char * const sections[], size_t *addrs, size_t *sizes) {
+static void locate_sections(
+                const PeSectionHeader section_table[],
+                size_t n_table,
+                struct PeSectionDescriptor **out_sections,
+                size_t *out_nsections,
+                bool in_memory) {
+
+        assert(section_table);
+        assert(out_sections);
+        assert(out_nsections);
+
+        *out_sections = xmalloc_multiply(sizeof **out_sections, n_table);
+        *out_nsections = 0;
+
+        for (const PeSectionHeader *section_hdr = section_table; section_hdr < section_table + n_table; ++section_hdr) {
+                struct PeSectionDescriptor *desc = *out_sections + *out_nsections++;
+                memcpy(desc->name, section_hdr->Name, sizeof section_hdr->Name);
+                desc->name[sizeof section_hdr->Name] = 0;
+                desc->offset = in_memory ? section_hdr->VirtualAddress : section_hdr->PointerToRawData;
+                desc->size = section_hdr->VirtualSize;
+        }
+
+        /* Sort sections
+         * NOTE: maybe bubblesort isn't ideal, but surely there aren't that many sections?
+         */
+        bool dirty;
+        do {
+                dirty = false;
+                for (size_t i = 1; i < *out_nsections; ++i) {
+                        int ret = strcmp8((*out_sections)[i-1].name,
+                                          (*out_sections)[i].name);
+                        if (ret > 0) {
+                                dirty = true;
+                                struct PeSectionDescriptor tmp = (*out_sections)[i-1];
+                                (*out_sections)[i-1] = (*out_sections)[i];
+                                (*out_sections)[i] = tmp;
+                        }
+                }
+        } while (dirty);
```

Comment:
please do not implement this manually here.

we have sort_pointer_array() already, which with very little work can be changed to sort any type (so that it comes closer to libc qsort()).

And then add a generic bsearch() please, to match that. Sooner or later we'll need this again after all, and it's much nicer to read then.

Reply:
Working on it.